src/test/java/com/springboot/intro/controller/CategoryControllerTest.java

```diff
@@ -0,0 +1,221 @@
+    @Test
+    @DisplayName("""
+            get three categories
+            """)
+    public void getAll_GetThreeCategories_Success() throws Exception {
+        List<CategoryResponseDto> expected = new ArrayList<>();
+        expected.add(new CategoryResponseDto(1L, "Fiction", "d1"));
+        expected.add(new CategoryResponseDto(2L, "Action", "d2"));
+        expected.add(new CategoryResponseDto(3L, "Thriller", "d3"));
+
+        MvcResult result = mockMvc.perform(get("/api/categories")
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        CategoryResponseDto[] actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                CategoryResponseDto[].class);
+        assertEquals(expected.size(), actual.length);
+        assertEquals(expected, Arrays.stream(actual).toList());
+    }
+
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @Test
+    @DisplayName("""
+            update category by id = 2
+            """)
+    public void updateCategory_ValidRequestDto_Success() throws Exception {
+        CategoryRequestDto categoryRequestDto = new CategoryRequestDto();
+        categoryRequestDto.setName("Poems");
+        categoryRequestDto.setDescription("about poems");
+
+        String jsonRequest = objectMapper.writeValueAsString(categoryRequestDto);
+
+        MvcResult result = mockMvc.perform(put("/api/categories/2")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        CategoryResponseDto actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                CategoryResponseDto.class);
+        assertEquals(2, actual.getId());
+        EqualsBuilder.reflectionEquals(categoryRequestDto, actual, "id");
+    }
+
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @Test
+    @DisplayName("""
+            update category by id = 1 with invalid request body
+            """)
+    public void updateCategory_InvalidRequestDto_BadRequest() throws Exception {
+        CategoryRequestDto categoryRequestDto = new CategoryRequestDto();
+        categoryRequestDto.setDescription("about poems");
+
+        String jsonRequest = objectMapper.writeValueAsString(categoryRequestDto);
+
+        MvcResult result = mockMvc.perform(put("/api/categories/2")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+                .andReturn();
+        ErrorValidationDto actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                ErrorValidationDto.class);
+        assertEquals("name can't be blank", actual.getErrors()[0]);
+        assertEquals(HttpStatus.BAD_REQUEST, actual.getStatus());
```

Comment:
let's separate our code into 3 sections split with a space:
* **Given** - essentially setting up our `expected` variables and conditions like `mock.when`
* **When** - performing actions that we are testing and getting `actual` variable
* **Then** - assertion of the given and actual values

And try to keep that separation across all tests

Reply:
more about it [here](https://martinfowler.com/bliki/GivenWhenThen.html)

Reply:
++